ts/src/pro/bingx.ts

```diff
@@ -224,6 +236,207 @@ export default class bingx extends bingxRest {
+            if (marketType === 'spot') {
+                throw new NotSupported (this.id + ' watchOrderBookForSymbols is not supported for spot markets yet');
+            }
+        }
+        for (let i = 0; i < symbolsAndTimeframes.length; i++) {
+            const symbolAndTimeframe = symbolsAndTimeframes[i];
+            const sym = symbolAndTimeframe[0];
+            const tf = symbolAndTimeframe[1];
+            const market = this.market (sym);
+            const rawTf = this.safeString (this.timeframes, tf, tf);
+            if (chosenTimeframe === undefined) {
+                chosenTimeframe = rawTf;
+            } else if (chosenTimeframe !== rawTf) {
+                throw new BadRequest (this.id + ' watchOHLCVForSymbols requires all timeframes to be the same');
+            }
+            const channelName = 'kline_' + chosenTimeframe;
+            messageHashes.push (this.getMessageHash ('ohlcv', channelName, market['symbol']));
+        }
+        const subscriptionHash = 'all@kline_' + chosenTimeframe;
+        const url = this.safeString (this.urls['api']['ws'], marketType);
+        const uuid = this.uuid ();
+        const request: Dict = {
+            'id': uuid,
+            'dataType': subscriptionHash,
+        };
+        if (marketType === 'swap') {
+            request['reqType'] = 'sub';
+        }
+        const subscriptionArgs: Dict = {
+            'symbols': symbolsAndTimeframes,
+            'limit': limit,
+            'interval': chosenTimeframe,
+            'params': params,
+        };
+        const [ symbol, timeframe, candles ] = await this.watchMultiple (url, messageHashes, request, [ subscriptionHash ], subscriptionArgs);
+        if (this.newUpdates) {
+            limit = candles.getLimit (symbol, limit);
+        }
+        const filtered = this.filterBySinceLimit (candles, since, limit, 0, true);
+        return this.createOHLCVObject (symbol, timeframe, filtered);
+    }
+
+    getOrderBookLimitByMarketType (marketType: string, limit: Int = undefined) {
+        if (limit === undefined) {
+            limit = 100;
+        } else {
+            if (marketType === 'swap' || marketType === 'future') {
+                limit = this.findNearestCeiling ([ 5, 10, 20, 50, 100 ], limit);
+            } else if (marketType === 'spot') {
+                limit = this.findNearestCeiling ([ 20, 100 ], limit);
+            }
+        }
+        return limit;
+    }
+
+    getMessageHash (unifiedChannel: string, exchangeChannel: string, symbol: Str) {
+        // sometimes exchangeChannel & unifiedChannel names might coincide
+        if (symbol !== undefined) {
+            // bidask:bookTicker@BTC/USDT
+            return unifiedChannel + '|' + exchangeChannel + '^' + symbol;
+        } else {
+            // bidasks@bookTicker
+            return unifiedChannel + 's' + '|' + exchangeChannel;
+        }
```

Comment:
tbh I think this function just makes things harder to understand, the messageHash should be more or less standard and easy to construct so I think we should do it inside the handle_x method explicitly.

Reply:
ouch :/ we have started doing this method across all other exchanges, which is the key for easier unification. what is harder there? it's in opposite, makes message-hash system easier to set/get in watch/handle methods... can you plz give it a second thought? it does not harm and works perfect across all exchanges where I had implemented them

Reply:
we will also get a big benefit from this method once we move to ED subscriptions

Reply:
> messageHash should be more or less standard ..

the goal of this method is exactly it to achieve "standard" for message hashing across lib (atm it's a total mess of different junctions)

Reply:
we already have this method across different exchanges (binance, kucoin, kraken,/futures)

Reply:
first, the messageHash should not contain exchange-specific names, second it's not clear how to use this method =m `unifiedCahnnel/exchangeChannel` etc this is confusing and does not simplify anything, is it that hard to make `"orderbook:" + symbol` ?

Reply:
> messageHash should not contain exchange-specific names

i'd say in opposite - from what I have collected while interacting different exchanges. it's not simply "orderbook: symbol", because orderbook endpoints (and many other endpoints) have sub-channel specific parameters, like "depth50/100/500" , "interval_100/500/1000" ... so the final messageHash should look like `"orderbook|${symbol}|depth100@interval500"` and then when data incomes, in handler, we have to find those depth & interval matches... and that's the reason I made that helper method, without it, it would have been a mess to manually construct and match those junctions individually on different places, and even to match the exact formatting (because separators need to be `|` or `^` or such symbols, which are not used by any exchanges, because in many cases exchange-specific strings can contain `@` or `:`, like  `BTC:USDT@ticker24` or etc.. so when splitting by `@` it would be unreliable.  while `|` and `^` is reliable.
and it's not only orderbook, it's  for candles o (i.e `ohlcv|${symbol}^5m`) or tickers (`ticker|${symbol}^rolling24` or etc) as  many endpoints have many different configurations 

I really don't understand how it makes harder, while it is really cumbersome to handle without that helper (which was the reason I had created them across different exchanges, and they simplified things quite much)

Reply:
we can remove exchange-specific parts in messageHash is inevitable (it might not always be used in 100% of cases, but in many cases it would be necessary that we supported that format, and also that format should be leaned to unification in long run)

Reply:
if you'd follow me on this (at least for a while) i promise you'd not regret

Reply:
>  so the final messageHash should look like "orderbook|${symbol}^depth100@interval500" and then when data incomes

this is not correct, we only store 1 ob per symbol, so it does not make sense to support multiple depths and intervals

> would have been a mess to manually construct and match those junctions individually on different places

this is the issue, messageHashes are meant to be very simple and not as you're describing them

> we can't remove exchange-specific parts in messageHash, because it has an inevitable need 
I would say that in 99% of the cases we can stick to the unified terms (orderbook/trades/myTrades/orders/etc), there's no need to bring exchange-specific terminology to the hash

Reply:
In short, the goal is to keep the code as simple as possible unless it's not possible, in this case we're overcomplicating it a lot for no reason